source/simulator/solver_schemes.cc

```diff
@@ -564,86 +576,113 @@ namespace aspect
-         * add each iteration an increasingly smaller solution update until the decreasing residual condition
-         * is met, or the line search iteration limit is reached.
-         */
-        LinearAlgebra::BlockVector backup_linearization_point = current_linearization_point;
-
-        double test_residual = 0;
-        double test_velocity_residual = 0;
-        double test_pressure_residual = 0;
-        double lambda = 1;
-        double alpha = 1e-4;
-        unsigned int line_search_iteration = 0;
-
-        /**
-         * Do the loop for the line search. Even when we
-         * don't do a line search we go into this loop
-         */
-        do
+        double test_residual = residual;
+        if (nonlinear_iteration == 0)
           {
-            current_linearization_point = backup_linearization_point;
-
-            LinearAlgebra::BlockVector search_direction = solution;
+            /**
+             * The first nonlinear iteration we are computing the whole system in a non-defect corrected Picard way,
+             * to make sure that the boundary conditions are correct in combination with correct initial guesses.
+             */
+            current_linearization_point.block(introspection.block_indices.velocities) = solution.block(introspection.block_indices.velocities);
+            current_linearization_point.block(introspection.block_indices.pressure) = solution.block(introspection.block_indices.pressure);
 
-            search_direction *= lambda;
+            residual = stokes_residual;
 
+            pcout << "      Relative nonlinear residual (total Newton system) after nonlinear iteration " << nonlinear_iteration+1
+                  << ": " << stokes_residual/initial_residual << ", norm of the rhs: " << stokes_residual << std::endl;
 
-            current_linearization_point.block(introspection.block_indices.pressure) += search_direction.block(introspection.block_indices.pressure);
-            current_linearization_point.block(introspection.block_indices.velocities) += search_direction.block(introspection.block_indices.velocities);
+          }
+        else
+          {
+            /**
+             * We may need to do a line search if the solution update doesn't decrease the norm of the rhs enough.
+             * This is done by adding the solution update to the current linearization point and then assembling
+             * the Newton right hand side. If the Newton residual has decreased enough by using this update,
+             * then we continue, otherwise we reset the current linearization point with the help of the backup and
+             * add each iteration an increasingly smaller solution update until the decreasing residual condition
+             * is met, or the line search iteration limit is reached.
+             */
+            LinearAlgebra::BlockVector backup_linearization_point = current_linearization_point;
```

Comment:
this assignment is not necessary

Reply:
I agree with you that the assignemnet should not be nessesary. The problem I face is that need to get the structure of the vector right. I tried `LinearAlgebra::BlockVector backup_linearization_point(introspection.index_sets.stokes_partitioning, mpi_communicator);`, but this gives me an error. How should I do this?
```
An error occurred in line <64> of file </home/fraters/deal.ii/deal.ii_8.5/dealii/source/lac/trilinos_block_vector.cc> in function
292:     dealii::TrilinosWrappers::MPI::BlockVector& dealii::TrilinosWrappers::MPI::BlockVector::operator=(const dealii::TrilinosWrappers::MPI::BlockVector&)
The violated condition was: 
    n_blocks() == 0 || n_blocks() == v.n_blocks()
Additional information: 
    Dimension 3 not equal to 2.
```

Reply:
This highlights another fact: your backup_linearization_point currently contains not only the stokes part but also temperature and all compositional fields. If you create it just for the Stokes part (like you wrote). You have to copy the individual blocks of velocity and pressure.

Reply:
hmm, you are right. But I noticed that I need an assignment anyway, so that the first current_linearisation_point is not set to zero every time. So the code now reads now.
```
LinearAlgebra::BlockVector backup_linearization_point(introspection.index_sets.stokes_partitioning, mpi_communicator);
backup_linearization_point.block(introspection.block_indices.pressure) = current_linearization_point.block(introspection.block_indices.pressure);
backup_linearization_point.block(introspection.block_indices.velocities) = current_linearization_point.block(introspection.block_indices.velocities);

```

Reply:
> But I noticed that I need an assignment anyway

Really? Inside the ``do`` you will be assigning it again and your are not reading it in-between or am I reading the diff incorrectly?

Reply:
The relevant code in that part now reads:
```
            LinearAlgebra::BlockVector backup_linearization_point(introspection.index_sets.stokes_partitioning, mpi_communicator);
            backup_linearization_point.block(introspection.block_indices.pressure) = current_linearization_point.block(introspection.block_indices.pressure);
            backup_linearization_point.block(introspection.block_indices.velocities) = current_linearization_point.block(introspection.block_indices.velocities);

...

            do
              {
                // Reset the current linearization point and the search direction
                current_linearization_point.block(introspection.block_indices.pressure) = backup_linearization_point.block(introspection.block_indices.pressure);
                current_linearization_point.block(introspection.block_indices.velocities) = backup_linearization_point.block(introspection.block_indices.velocities);
...
```
So for the first loop, the backup_linearization_point needs to have the value of current_linearization_point. After this point the backup_linearization_point is never reassigned.